nixos/nginx: use sandboxing mode

Izorkin opened the pull request:
###### Motivation for this change
PR based on https://github.com/NixOS/nixpkgs/pull/60646
Enable run nginx web service  in sandboxing mode.

cc @dasJ @flokli @aanderse @Mic92

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

Izorkin committed:
nixos/nginx: enable sandboxing

Izorkin committed:
nixos/nginx: add release notes

dasJ commented:
A fully up-to-date list of hardening options our sandbox module applies:
```
            # Filesystem stuff
            # TODO Force for phpfpm
            ProtectSystem = "strict"; # Prevent writing to most of /
            ProtectHome = true; # Prevent accessing /home and /root
            PrivateTmp = true; # Give an own directory under /tmp
            PrivateDevices = true; # Deny access to most of /dev
            ProtectKernelTunables = true; # Protect some parts of /sys
            ProtectControlGroups = true; # Remount cgroups read-only
            RestrictSUIDSGID = true; # Prevent creating SETUID/SETGID files
            PrivateMounts = true; # Give an own mount namespace

            # Capabilities
            CapabilityBoundingSet = ""; # Allow no capabilities at all
            NoNewPrivileges = true; # Disallow getting more capabilities. This is also implied by other options.

            # Kernel stuff
            ProtectKernelModules = true; # Prevent loading of kernel modules
            SystemCallArchitectures = "native"; # Usually no need to disable this
            ProtectKernelLogs = true; # Prevent access to kernel logs
            ProtectClock = true; # Prevent setting the RTC

            # Networking
            RestrictAddressFamilies = ""; # Example: "AF_UNIX AF_INET AF_INET6"
            PrivateNetwork = true; # Isolate the entire network

            # Misc
            LockPersonality = true; # Prevent change of the personality
            ProtectHostname = true; # Give an own UTS namespace
            RestrictRealtime = true; # Prevent switching to RT scheduling
            MemoryDenyWriteExecute = true; # Maybe disable this for interpreters like python
            PrivateUsers = true; # If anything randomly breaks, it's mostly because of this
```

I understand some of these are not useful at all (like `PrivateNetwork`), but maybe you find some of these useful.

Izorkin commented:
With `PrivateUsers = true;` error:
```
nginx.service: Failed to set up user namespacing: No space left on device
nginx.service: Failed at step USER spawning /nix/store/snsli9q204d8xhiwx8vr5faq5snd153c-unit-script-nginx-pre-start: No space left on device
```
With `ProtectClock = true;` and `ProtectKernelLogs = true;` warning:

```
systemd[1]: /nix/store/bsjqsa1jpy2kgxr5h0y177hcn7wpv9df-unit-nginx.service/nginx.service:26: Unknown key name 'ProtectClock' in section 'Service', ignoring.
systemd[1]: /nix/store/bsjqsa1jpy2kgxr5h0y177hcn7wpv9df-unit-nginx.service/nginx.service:30: Unknown key name 'ProtectKernelLogs' in section 'Service', ignoring.
```
With `CapabilityBoundingSet = "";`
```
nginx.service: Failed to apply ambient capabilities (before UID change): Operation not permitted
nginx.service: Failed at step CAPABILITIES spawning /nix/store/snsli9q204d8xhiwx8vr5faq5snd153c-unit-script-nginx-pre-start: Operation not permitted
```

dasJ commented:
Our nginx module uses these for the capabilities:
```
      CapabilityBoundingSet = [ "CAP_NET_BIND_SERVICE" "CAP_SYS_RESOURCE" ];
      AmbientCapabilities = [ "CAP_NET_BIND_SERVICE" "CAP_SYS_RESOURCE" ];
```

Izorkin commented:
Updated.

dasJ commented:
Are you sure it needs AF_NETLINK? It's weird because it's lacking from our module, but is then allowed in AppArmor later on :thinking: @ajs124 any idea?

ajs124 commented:
I have no recollection of anything related to this, so you might as well dig through the code yourself @dasJ.

dasJ commented:
Also it's worth mentioning in the changelog that `MemoryDenyWriteExecute` will probably break the nginx lua module :weary:

Izorkin commented:
Remove this option 

> Also it's worth mentioning in the changelog that `MemoryDenyWriteExecute` will probably break the nginx lua module 😩

Remove the option `MemoryDenyWriteExecute`?

dasJ commented:
It's a useful way of mitigating more severe problems when an attacker has already gained some access, so I'd keep it but people using the lua module should know that it breaks the module.

Izorkin commented:
Add this wariant?
`MemoryDenyWriteExecute = mkDefault true;`

dasJ commented:
Probably for the best, yes :)

Izorkin commented:
With `MemoryDenyWriteExecute = true;` nginx not worked, time-out open test page.

Izorkin commented:
Not worked with `pkgs.nginxModules.pagespeed`

Izorkin commented:
@dasJ need backport to 20-03?

flokli commented:
This can't be backported to 20.03, which is almost released already.

I'm also a bit afraid about the things some of these options might break - do any other distros already have a nginx unit with some sandboxing options we could take some inspiration from?

Ma27 commented:
@GrahamcOfBorg build nginx

flokli commented on nixos/doc/manual/release-notes/rl-2009.xml:
```diff
@@ -212,7 +212,20 @@ environment.systemPackages = [
-
+   <listitem>
+     <para>
+       Now nginx web server started in sandbox mode. By default, recording is
+       allowed in <literal>/var/spool/nginx</literal>. To allow writing to other folders,
```
```suggestion
       Now nginx web server started in sandbox mode. By default, write access to
       <literal>/var/spool/nginx</literal> is allowed. To allow writing to other folders,
```

flokli replied:
Also, this is not `/var/spool/nginx`, but whatever `cfg.stateDir` is currently pointing to.

IIRC, we wanted to move this to `/var/lib/nginx` at some point of time.

Izorkin replied:
> IIRC, we wanted to move this to `/var/lib/nginx` at some point of time.

such a variant?
```
commit fbb86293947d326c4af08a214ecf41da40a80a36
Refs: [test-nginx], 18.09-beta-70407-gfbb86293947
Author:     Izorkin <izorkin@elven.pw>
AuthorDate: Sun Apr 19 17:50:11 2020 +0300
Commit:     Izorkin <izorkin@elven.pw>
CommitDate: Sun Apr 19 17:51:22 2020 +0300

    nginx: change logs path
---
 pkgs/servers/http/nginx/generic.nix                    | 12 +++++++++++-
 pkgs/servers/http/nginx/nix-skip-check-logs-path.patch | 27 +++++++++++++++++++++++++++
 2 files changed, 38 insertions(+), 1 deletion(-)

diff --git a/pkgs/servers/http/nginx/generic.nix b/pkgs/servers/http/nginx/generic.nix
index 67a914b6a98..63408be6a21 100644
--- a/pkgs/servers/http/nginx/generic.nix
+++ b/pkgs/servers/http/nginx/generic.nix
@@ -68,6 +68,14 @@ stdenv.mkDerivation {
     "--with-http_stub_status_module"
     "--with-threads"
     "--with-pcre-jit"
+    "--http-log-path=/var/log/nginx/access.log"
+    "--error-log-path=/var/log/nginx/error.log"
+    "--pid-path=/var/log/nginx/nginx.pid"
+    "--http-client-body-temp-path=/var/cache/nginx/client_body"
+    "--http-proxy-temp-path=/var/cache/nginx/proxy"
+    "--http-fastcgi-temp-path=/var/cache/nginx/fastcgi"
+    "--http-uwsgi-temp-path=/var/cache/nginx/uwsgi"
+    "--http-scgi-temp-path=/var/cache/nginx/scgi"
   ] ++ optionals withDebug [
     "--with-debug"
   ] ++ optionals withStream [
@@ -105,7 +113,9 @@ stdenv.mkDerivation {
       preInstall = ''
         export nixStoreDir="$NIX_STORE" nixStoreDirLen="''${#NIX_STORE}"
       '';
-    }) ++ optionals (stdenv.hostPlatform != stdenv.buildPlatform) [
+    }) ++ [
+      ./nix-skip-check-logs-path.patch
+    ] ++ optionals (stdenv.hostPlatform != stdenv.buildPlatform) [
       (fetchpatch {
         url = "https://raw.githubusercontent.com/openwrt/packages/master/net/nginx/patches/102-sizeof_test_fix.patch";
         sha256 = "0i2k30ac8d7inj9l6bl0684kjglam2f68z8lf3xggcc2i5wzhh8a";
diff --git a/pkgs/servers/http/nginx/nix-skip-check-logs-path.patch b/pkgs/servers/http/nginx/nix-skip-check-logs-path.patch
new file mode 100644
index 00000000000..a823660cc32
--- /dev/null
+++ b/pkgs/servers/http/nginx/nix-skip-check-logs-path.patch
@@ -0,0 +1,27 @@
+diff --git a/auto/install b/auto/install
+index d884487..dccc411 100644
+--- a/auto/install
++++ b/auto/install
+@@ -148,12 +148,6 @@ install:  build $NGX_INSTALL_PERL_MODULES
+               || cp conf/nginx.conf '\$(DESTDIR)$NGX_CONF_PATH'
+       cp conf/nginx.conf '\$(DESTDIR)$NGX_CONF_PREFIX/nginx.conf.default'
+
+-      test -d '\$(DESTDIR)`dirname "$NGX_PID_PATH"`' \\
+-              || mkdir -p '\$(DESTDIR)`dirname "$NGX_PID_PATH"`'
+-
+-      test -d '\$(DESTDIR)`dirname "$NGX_HTTP_LOG_PATH"`' \\
+-              || mkdir -p '\$(DESTDIR)`dirname "$NGX_HTTP_LOG_PATH"`'
+-
+       test -d '\$(DESTDIR)$NGX_PREFIX/html' \\
+               || cp -R $NGX_HTML '\$(DESTDIR)$NGX_PREFIX'
+ END
+@@ -161,9 +155,6 @@ END
+
+ if test -n "$NGX_ERROR_LOG_PATH"; then
+     cat << END                                                >> $NGX_MAKEFILE
+-
+-      test -d '\$(DESTDIR)`dirname "$NGX_ERROR_LOG_PATH"`' \\
+-              || mkdir -p '\$(DESTDIR)`dirname "$NGX_ERROR_LOG_PATH"`'
+ END
+
+ fi
```

flokli replied:
We might want to do a separate PR in nixpkgs changing paths to more standard locations before starting with sandboxing.

flokli replied:
Also, when doing so, pleas see if we can fix the underlying issue `nix-skip-check-logs-path.patch` is trying to workaround.

Izorkin replied:
> Also, when doing so, pleas see if we can fix the underlying issue `nix-skip-check-logs-path.patch` is trying to workaround.

Variant - build without sandox.

flokli commented on nixos/doc/manual/release-notes/rl-2009.xml:
```diff
@@ -212,7 +212,20 @@ environment.systemPackages = [
+       <programlisting>
+systemd.services.nginx.serviceConfig.ReadWritePaths = [ "/var/www" ];
+       </programlisting>
+       Breaks the nginx lua and pagespeed modules. To correct work need disable <literal>MemoryDenyWriteExecute</literal>
```
Can we test if these modules are enabled, and only enable `MemoryDenyWriteExecute` when they're not enabled?

Izorkin replied:
I don't know how to implement

Izorkin replied:
@flokli simple worked variant 
```
  nixpkgs.overlays = [
    (self: super: {
      php = pkgs.php72;
      phpPackages = pkgs.php72Packages;
      nginx = super.nginx.override {
        withDebug = false;
        withStream = false;
        withMail = false;
        modules = [
          pkgs.nginxModules.lua
        ];
      };
    })
  ];
```
and
```
MemoryDenyWriteExecute = (if pkgs.nginx.modules == [ pkgs.nginxModules.lua ] then false else true);
```
If change to 
```
        modules = [
          pkgs.nginxModules.lua
          pkgs.nginxModules.rtmp
        ];
```
Not worked.
What other method is there?

Izorkin replied:
Added test to check are enabled modules.

flokli commented on nixos/modules/services/web-servers/nginx/default.nix:
```diff
@@ -708,8 +708,30 @@ in
+        ReadWritePaths = [ cfg.stateDir ];
         # Capabilities
         AmbientCapabilities = [ "CAP_NET_BIND_SERVICE" "CAP_SYS_RESOURCE" ];
+        CapabilityBoundingSet = [ "CAP_NET_BIND_SERVICE" "CAP_SYS_RESOURCE" ];
+        # Security
+        NoNewPrivileges = true;
+        # Sandboxing
+        ProtectSystem = "strict";
+        ProtectHome = true;
+        PrivateTmp = true;
+        PrivateDevices = true;
+        ProtectHostname = true;
+        ProtectKernelTunables = true;
+        ProtectKernelModules = true;
+        ProtectControlGroups = true;
+        RestrictAddressFamilies = [ "AF_UNIX" "AF_INET" "AF_INET6" ];
+        LockPersonality = true;
+        MemoryDenyWriteExecute = mkDefault true;
+        RestrictRealtime = true;
+        RestrictSUIDSGID = true;
+        PrivateMounts = true;
+        # System Call Filtering
+        SystemCallArchitectures = "native";
```
Again, I'd feel wayy more comfortable if we'd have more tests covering standard usecases, and see they're not broken.

Izorkin replied:
Need create test to check sandbox mode with perl /lua scripts?

flokli replied:
yes. We need more tests to ensure things are not broken.

Izorkin committed:
nixos/tests: fix nginx-pubhtml test

flokli commented:
I don't really think this is anywhere near ready. We need to have way more testing in place before making such a drastical change on the nginx module.

Izorkin committed:
nixos/tests: add nginx-sandbox test

Izorkin commented:
@GrahamcOfBorg test nginx-sandbox

Izorkin committed:
nginxModules: add option allowMemoryWriteExecute

The allowMemoryWriteExecute option is required to checking enabled nginxModules
and disable the nginx sandbox mode MemoryDenyWriteExecute.

Izorkin commented:
@GrahamcOfBorg test nginx
@GrahamcOfBorg test nginx-sandbox

aanderse commented:
I'm not a `nginx` user so I could certainly be wrong, but I worry these changes might be excessively opinionated for a linux distro to push onto users for a web server. Generally a web server is a bit too varied in what it can do that I would have thought applying changes like these is a bit heavy handed. I would suggest writing a section in the NixOS manual on hardening services and then add a link in the `nginx` section of the manual suggesting these changes.

Izorkin commented:
@aanderse i don’t think I can write such a section :(

Mic92 commented on pkgs/servers/http/nginx/modules.nix:
```diff
@@ -246,6 +267,7 @@ in
     in {
       src = ngx_pagespeed;
       inputs = [ pkgs.zlib pkgs.libuuid ]; # psol deps
+      allowMemoryWriteExecute = true;
```
I think this could default to `true` so it is only disabled in the other cases.

Izorkin replied:
If add `allowMemoryWriteExecute = true;` only `lua` modules
```
building all machine configurations...
error: attribute 'allowMemoryWriteExecute' missing, at /home/lf/works/src-nix/nixpkgs/nixos/modules/services/web-servers/nginx/default.nix:727:55

```

Mic92 replied:
see https://github.com/NixOS/nixpkgs/pull/85567#discussion_r423031088 on how to implement this.

Izorkin replied:
Fixed.

Mic92 commented on nixos/modules/services/web-servers/nginx/default.nix:
```diff
@@ -710,6 +710,26 @@ in
+        ProtectControlGroups = true;
+        RestrictAddressFamilies = [ "AF_UNIX" "AF_INET" "AF_INET6" ];
+        LockPersonality = true;
+        MemoryDenyWriteExecute = !(builtins.any (mod: mod.allowMemoryWriteExecute) pkgs.nginx.modules);
```
This allows to provide a default if the module does not has one:

```suggestion
        MemoryDenyWriteExecute = !(builtins.any (mod: (mod.allowMemoryWriteExecute or false)) pkgs.nginx.modules);
```

Izorkin replied:
Fixed.

Mic92 replied:
Did you forget to push the changes?

Izorkin replied:
Just updated.

Mic92 commented on nixos/doc/manual/release-notes/rl-2009.xml:
```diff
@@ -235,7 +235,20 @@ php.override {
-
+   <listitem>
+     <para>
+       Now nginx web server started in sandbox mode. By default, write access to
```
```suggestion
       Nginx web server is now started with additional sandbox/hardening options. By default, write access to
```

Izorkin replied:
Thanks, fixed.

Mic92 commented on nixos/doc/manual/release-notes/rl-2009.xml:
```diff
@@ -235,7 +235,20 @@ php.override {
+       <programlisting>
+systemd.services.nginx.serviceConfig.ReadWritePaths = [ "/var/www" ];
+       </programlisting>
+       Breaks the nginx lua and pagespeed modules. To correct work need disable <literal>MemoryDenyWriteExecute</literal>
```
This line can be dropped now.

Izorkin commented:
@GrahamcOfBorg eval

flokli commented:
After some thinking, I agree with @aanderse and https://github.com/NixOS/nixpkgs/pull/85567#issuecomment-626662085.

We shouldn't push something as drastical as this to users without some broader community discussion on the topic.

Even if we decide to do this, it should probably be done in a much more gradual approach - providing an option to enable sandboxing, encouraging users to test it, and only default it (if at all) when it has seen enough feedback. All together with a lot of communication, release notes etc.

Izorkin committed:
nixos/nginx: add option enableSandbox

Izorkin commented:
@flokli @aanderse @Mic92 added option to enable sandbox mode. Sandbox mode is disabled by default.

Izorkin commented:
@GrahamcOfBorg test nginx
@GrahamcOfBorg test nginx-sandbox

Mic92 commented on nixos/modules/services/web-servers/nginx/default.nix:
```diff
@@ -710,6 +718,27 @@ in
+      } // optionalAttrs cfg.enableSandbox {
+        # Sandboxing
+        ProtectSystem = "strict";
+        ProtectHome = mkDefault true;
```
How about moving `ProtectSystem`, `ProtectHome` and `MemoryDenyWriteExecute` to `enableSandbox` and move other options above the conditional? I think those 3 options  are the controversial ones that could break with peoples setup, where we need more feedback.

Izorkin replied:
This variant?
```
        # Security
        NoNewPrivileges = true;
        # Sandboxing
        PrivateTmp = true;
        PrivateDevices = true;
        ProtectHostname = true;
        ProtectKernelTunables = true;
        ProtectKernelModules = true;
        ProtectControlGroups = true;
        RestrictAddressFamilies = [ "AF_UNIX" "AF_INET" "AF_INET6" ];
        LockPersonality = true;
        RestrictRealtime = true;
        RestrictSUIDSGID = true;
        PrivateMounts = true;
        # System Call Filtering
        SystemCallArchitectures = "native";
      } // optionalAttrs cfg.enableSandbox {
        ProtectSystem = "strict";
        ProtectHome = mkDefault true;
        MemoryDenyWriteExecute = !(builtins.any (mod: (mod.allowMemoryWriteExecute or false)) pkgs.nginx.modules);
      };

```

Mic92 replied:
Yes. This looks good to me.

Izorkin replied:
In this variant, it will not work to disable the options `PrivateTmp` and etc.

Mic92 replied:
Right. Maybe `PrivateTmp` should be also in the `sandbox` section.

Izorkin replied:
Or leave the current variant, or rename `cfg.enableSandbox` to `cfg.enableSandboxStrict`.

Izorkin replied:
I prefer the current variant.

nixbitcoin commented on nixos/modules/services/web-servers/nginx/default.nix:
```diff
@@ -710,6 +718,27 @@ in
+        RestrictSUIDSGID = true;
+        PrivateMounts = true;
+        # System Call Filtering
+        SystemCallArchitectures = "native";
```
Consider implementing a `SystemCallFilter` with sane system calls like  https://github.com/fort-nix/nix-bitcoin/blob/4b5d0947facef3464aa868038f6365fa90e571ae/modules/nix-bitcoin-services.nix#L28

Mic92 replied:
I would leave this out of now since we don't know what our modules do. This could be added later.

Izorkin replied:
Not enough knowledge to define filters.

Mic92 replied:
Yes. Leave this for now.

nixbitcoin commented:
I integrated the earlier module from #60646 in https://github.com/nixbitcoin/nix-bitcoin/tree/harden-nginx and never had any problems. I would be very happy to see `services.nginx.enableSandbox` implemented in nixpkgs.

nixbitcoin commented:
Consider adding `RemoveIPC` and `PrivateUsers` systemd hardening options.

Izorkin commented on nixos/modules/services/web-servers/nginx/default.nix:
```diff
@@ -463,6 +463,14 @@ in
       };
 
+      enableSandbox = mkOption {
+        default = false;
```
Can make `default = true;`, and change to release  `default = false;` ?

Mic92 replied:
No. Just keep it as it is right now and once we gained some confidence and feedback, we change it to true or remove the option completely.

Izorkin commented:
Added `RemoveIPC`

Izorkin commented:
@GrahamcOfBorg test nginx
@GrahamcOfBorg test nginx-sandbox

Mic92 commented on nixos/modules/services/web-servers/nginx/default.nix:
```diff
@@ -710,6 +718,28 @@ in
+        MemoryDenyWriteExecute = !(builtins.any (mod: (mod.allowMemoryWriteExecute or false)) pkgs.nginx.modules);
+        RestrictRealtime = true;
+        RestrictSUIDSGID = true;
+        RemoveIPC = true;
```
I think this might break some features in nginx. I see that it uses `shmget` and `shmat` internally. I would skip this flag.

Mic92 replied:
https://github.com/nginx/nginx/blob/4bf4650f2f10f7bbacfe7a33da744f18951d416d/src/os/unix/ngx_shmem.c#L102

Izorkin replied:
Removed.

nh2 commented on nixos/doc/manual/release-notes/rl-2009.xml:
```diff
@@ -235,7 +235,16 @@ php.override {
+   <listitem>
+     <para>
+       Add option <literal>services.nginx.enableSandbox</literal> to starting Nginx web server with additional sandbox/hardening options.
+       By default, write access to <literal>services.nginx.stateDir</literal> is allowed. To allow writing to other folders,
```
I think that the release notes should also have mentioned that due to the added `ProtectHome = mkDefault true;`, nginx can no longer serve files from users' home directories.

It's not uncommon to serve subdirectories of home directories publicly (e.g. `/home/*/public_html/`), and this change breaks that without any mention.

nh2 commented:
> I think that the release notes should also have mentioned that due to the added `ProtectHome = mkDefault true;`, nginx can no longer serve files from users' home directories.

I made PR #103147 for that.

Needs 20.09 backport.
